### PR TITLE
Add skipLines in FileIO.ReadableFile.readFullyAsUTF8String

### LINKED ISSUE #17990: Skip header row from csv
Add functionality to skip header rows when reading from a csv file.

http://stackoverflow.com/questions/28450554/skipping-header-rows-is-it-possible-with-cloud-dataflow

Imported from Jira [BEAM-123](https://issues.apache.org/jira/browse/BEAM-123). Original Jira may contain additional context.
Reported by: davin.pidoto.

### PR DESCRIPTION
**Pull request by gudfhr95**
Fixes #17990

**Changes**
- Add `skipLines` in `FileIO.ReadableFile.readFullyAsUTF8String`
- Add `skipLines` in `FileIO.ReadableFile.readFullyAsBytes`
- Refactored `FileIOTest`to `Enclosed.class`

**Related PR**
- https://github.com/apache/beam/pull/28502

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by gudfhr95** [comment]
Run Java_PVR_Flink_Docker PreCommit

**Comment by gudfhr95** [comment]
Run Java_IOs_Direct PreCommit

**Comment by liferoad** [comment]
R: @shunping-google

**Comment by gudfhr95** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by gudfhr95** [comment]
Run Java_Pulsar_IO_Direct PreCommit

**Comment by gudfhr95** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by gudfhr95** [comment]
Run Java PreCommit

**Comment by gudfhr95** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by gudfhr95** [comment]
Hi, @damondouglas.
I finished implementing what you suggested.
I'm looking forward for your review.
Thank you.

**Comment by damccorm** [comment]
@damondouglas or @shunping-google could you please take a look at this one?

**Comment by shunping-google** [comment]
Hi [gudfhr95](https://github.com/gudfhr95), I am reviewing your change today and will get back to you shortly. Sorry about the delay.

**Comment by shunping-google** [comment]
[gudfhr95](https://github.com/gudfhr95),

Thanks again for submitting your PRs for review! I spent some time discussing with the previous reviewer (@damondouglas) about this PR (https://github.com/apache/beam/pull/28728) offline and below is our opinion.

While there are merits in your changes, we also see certain problems that we will have to address:
* It modifies the signature of the public function readFullyAsBytes(), which will inevitably break the backward compatibility of existing code. This is something we want to avoid.
* It may not work on large files since it has to put all file content in the memory.
* It only considers "\n" as the line delimiter, while there could be more out there.

In fact, I think FileIO is not the best place for this change. In the design of Beam IO, FileIO should be mainly focusing on getting a readable/writable channel ready. It should not have any assumption on what's inside the file. If you put the header skipping feature in FileIO, you are assuming it is a text-like file that has line breaks. 

We prefer putting the change in a class derived from FileIO. Just like your original PR (https://github.com/apache/beam/pull/28502), TextIO will be the best choice as it is for text files and it has already handled different types of delimiters. However, this PR also introduced some breaking changes to the public functions. Moreover, it will not scale well because it needs to disable isSplittable() when the header skipping feature is used.

In summary, we see problems in both PRs and it will be great if we can 

1. implement the changes in TextIO,

1. keep the capability of parallel processing for a large file, and

1. make sure there no breaking changes (instead of changing the signature of a public function, try to overload it).

There is another PR (https://github.com/apache/beam/pull/29202) submitted recently on implementing this feature. You can refer to it for some ideas.

Thanks,

Shunping
